File: python/dolfin/generation.py
# -*- coding: utf-8 -*-
# Copyright (C) 2017-2019 Chris N. Richardson and Michal Habera
#
# This file is part of DOLFIN (https://www.fenicsproject.org)
#
# SPDX-License-Identifier:    LGPL-3.0-or-later
"""Simple mesh generation module"""

import typing
import numpy
from dolfin import cpp, fem

__all__ = ["IntervalMesh", "UnitIntervalMesh",
           "RectangleMesh", "UnitSquareMesh",
           "BoxMesh", "UnitCubeMesh"]


def IntervalMesh(comm,
                 nx: int,
                 points: list,
                 ghost_mode=cpp.mesh.GhostMode.none):
    """Create an interval mesh

    Parameters
    ----------
    comm
        MPI communicator
    nx
        Number of cells
    point
        Coordinates of the end points
    ghost_mode

    Note
    ----
    Coordinate mapping is not attached
    """
    return cpp.generation.IntervalMesh.create(comm, nx, points, ghost_mode)


def UnitIntervalMesh(comm, nx, ghost_mode=cpp.mesh.GhostMode.none):
    """Create a mesh on the unit interval with coordinate mapping attached

    Parameters
    ----------
    comm
        MPI communicator
    nx
        Number of cells

    """
    mesh = IntervalMesh(comm, nx, [0.0, 1.0], ghost_mode)
    mesh.geometry.coord_mapping = fem.create_coordinate_map(mesh)
    return mesh


def RectangleMesh(comm,
                  points: typing.List[numpy.array],
                  n: list,
                  cell_type=cpp.mesh.CellType.triangle,
                  ghost_mode=cpp.mesh.GhostMode.none,
                  diagonal: str = "right"):
    """Create rectangle mesh

    Parameters
    ----------
    comm
        MPI communicator
    points
        List of `Points` representing vertices
    n
        List of number of cells in each direction
    diagonal
        Direction of diagonal

    Note
    ----
    Coordinate mapping is not attached

    """
    return cpp.generation.RectangleMesh.create(comm, points, n, cell_type, ghost_mode, diagonal)


def UnitSquareMesh(comm,
                   nx,
                   ny,
                   cell_type=cpp.mesh.CellType.triangle,
                   ghost_mode=cpp.mesh.GhostMode.none,
                   diagonal="right"):
    """Create a mesh of a unit square with coordinate mapping attached

    Parameters
    ----------
    comm
        MPI communicator
    nx
        Number of cells in "x" direction
    ny
        Number of cells in "y" direction
    diagonal
        Direction of diagonal

    """
    print("AAAAAAAAA")
    mesh = RectangleMesh(comm, [numpy.array([0.0, 0.0, 0.0]),
                                numpy.array([1.0, 1.0, 0.0])],
                         [nx, ny], cell_type, ghost_mode, diagonal)
    print("BBBBBBBBB")
    mesh.geometry.coord_mapping = fem.create_coordinate_map(mesh)
    return mesh


def BoxMesh(comm,
            points: typing.List[numpy.array],
            n: list,
            cell_type=cpp.mesh.CellType.tetrahedron,
            ghost_mode=cpp.mesh.GhostMode.none):
    """Create box mesh

    Parameters
    ----------
    comm
        MPI communicator
    points
        List of points representing vertices
    n
        List of cells in each direction

    Note
    ----
    Coordinate mapping is not attached
    """
    return cpp.generation.BoxMesh.create(comm, points, n, cell_type, ghost_mode)


def UnitCubeMesh(comm,
                 nx,
                 ny,
                 nz,
                 cell_type=cpp.mesh.CellType.tetrahedron,
                 ghost_mode=cpp.mesh.GhostMode.none):
    """Create a mesh of a unit cube with coordinate mapping attached

    Parameters
    ----------
    comm
        MPI communicator
    nx
        Number of cells in "x" direction
    ny
        Number of cells in "y" direction
    nz
        Number of cells in "z" direction

    """
    mesh = BoxMesh(comm, [numpy.array([0.0, 0.0, 0.0]),
                          numpy.array([1.0, 1.0, 1.0])],
                   [nx, ny, nz], cell_type, ghost_mode)
    mesh.geometry.coord_mapping = fem.create_coordinate_map(mesh)
    return mesh

File: python/dolfinx/wrappers/generation.cpp
// Copyright (C) 2017 Chris N. Richardson and Garth N. Wells
//
// This file is part of DOLFINx (https://www.fenicsproject.org)
//
// SPDX-License-Identifier:    LGPL-3.0-or-later

#include "MPICommWrapper.h"
#include "caster_mpi.h"
#include <array>
#include <dolfinx/fem/CoordinateElement.h>
#include <dolfinx/generation/BoxMesh.h>
#include <dolfinx/generation/IntervalMesh.h>
#include <dolfinx/generation/RectangleMesh.h>
#include <pybind11/functional.h>
#include <pybind11/numpy.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <string>

namespace py = pybind11;

namespace
{
using PythonCellPartitionFunction
    = std::function<dolfinx::graph::AdjacencyList<std::int32_t>(
        dolfinx_wrappers::MPICommWrapper, int, int,
        const dolfinx::graph::AdjacencyList<std::int64_t>&,
        dolfinx::mesh::GhostMode)>;

using CppCellPartitionFunction
    = std::function<dolfinx::graph::AdjacencyList<std::int32_t>(
        MPI_Comm, int, int, const dolfinx::graph::AdjacencyList<std::int64_t>&,
        dolfinx::mesh::GhostMode)>;

CppCellPartitionFunction
create_cell_partitioner_cpp(const PythonCellPartitionFunction& partitioner)
{
  return [partitioner](MPI_Comm comm, int n, int tdim,
                       const dolfinx::graph::AdjacencyList<std::int64_t>& cells,
                       dolfinx::mesh::GhostMode ghost_mode)
  {
    return partitioner(dolfinx_wrappers::MPICommWrapper(comm), n, tdim, cells,
                       ghost_mode);
  };
}

} // namespace

namespace dolfinx_wrappers
{

void generation(py::module& m)
{
  py::enum_<dolfinx::generation::DiagonalType>(m, "DiagonalType")
      .value("left", dolfinx::generation::DiagonalType::left)
      .value("right", dolfinx::generation::DiagonalType::right)
      .value("crossed", dolfinx::generation::DiagonalType::crossed)
      .value("left_right", dolfinx::generation::DiagonalType::left_right)
      .value("right_left", dolfinx::generation::DiagonalType::right_left);

  m.def(
      "create_interval_mesh",
      [](const MPICommWrapper comm, std::size_t n, std::array<double, 2> p,
         dolfinx::mesh::GhostMode ghost_mode,
         const PythonCellPartitionFunction& partitioner)
      {
        return dolfinx::generation::IntervalMesh::create(
            comm.get(), n, p, ghost_mode,
            create_cell_partitioner_cpp(partitioner));
      },
      py::arg("comm"), py::arg("n"), py::arg("p"), py::arg("ghost_mode"),
      py::arg("partitioner"));

  m.def(
      "create_rectangle_mesh",
      [](const MPICommWrapper comm,
         const std::array<std::array<double, 3>, 2>& p,
         std::array<std::size_t, 2> n, dolfinx::mesh::CellType celltype,
         dolfinx::mesh::GhostMode ghost_mode,
         const PythonCellPartitionFunction& partitioner,
         dolfinx::generation::DiagonalType diagonal)
      {
        return dolfinx::generation::RectangleMesh::create(
            comm.get(), p, n, celltype, ghost_mode,
            create_cell_partitioner_cpp(partitioner), diagonal);
      },
      py::arg("comm"), py::arg("p"), py::arg("n"), py::arg("celltype"),
      py::arg("ghost_mode"), py::arg("partitioner"), py::arg("diagonal"));

  m.def(
      "create_box_mesh",
      [](const MPICommWrapper comm,
         const std::array<std::array<double, 3>, 2>& p,
         std::array<std::size_t, 3> n, dolfinx::mesh::CellType celltype,
         dolfinx::mesh::GhostMode ghost_mode,
         const PythonCellPartitionFunction& partitioner)
      {
        return dolfinx::generation::BoxMesh::create(
            comm.get(), p, n, celltype, ghost_mode,
            create_cell_partitioner_cpp(partitioner));
      },
      py::arg("comm"), py::arg("p"), py::arg("n"), py::arg("celltype"),
      py::arg("ghost_mode"), py::arg("partitioner"));
}
} // namespace dolfinx_wrappers
